### PR TITLE
Add solver structure

### PR DESCRIPTION
**Pull request by tmigot**
Add a `PercivalSolver` and use in-place NLPModel API in main function.

This is a first step remains:
- add subproblem in the structure
- add solverstructure of the subproblems
- test the reusability
- Tailor `solve!` to the workflow with `Val(ineq)`, `Val(tron)`.

### CODE REVIEW

File: src/method.jl [hunk 1]
```diff
@@ -70,8 +75,36 @@ Implementation of an augmented Lagrangian method. The following keyword paramete
 - inity: Initial values of the Lagrangian multipliers
 - subsolver_kwargs: subsolver keyword arguments as a dictionary
 """
-function percival(
+mutable struct PercivalSolver{T, V}
```
Comment (amontoison) [inline src/method.jl:78]: Why do you use `T`?

File: src/method.jl [hunk 2]
```diff
@@ -70,8 +75,36 @@ Implementation of an augmented Lagrangian method. The following keyword paramete
...
+  x::V
+  gx::V
+  gL::V
+  gp::V
+  Jtv::V
+end
+
+function PercivalSolver(nlp::M) where {T, V, M <: AbstractNLPModel{T, V}}
```
Comment (amontoison) [inline src/method.jl:86]:
```suggestion
function PercivalSolver(nlp::AbstractNLPModel{T, V}) where {T, V}
```

File: src/method.jl [hunk 3]
```diff
@@ -97,12 +130,14 @@ function percival(
...
 
-  x = copy(nlp.meta.x0)
-  x .= max.(nlp.meta.lvar, min.(x, nlp.meta.uvar))
+  x = solver.x
+  gx = solver.gx
+  x .= max.(nlp.meta.lvar, min.(nlp.meta.x0, nlp.meta.uvar))
 
-  gp = zeros(T, nlp.meta.nvar)
+  gp = solver.gp
+  gp .= zeros(T, nlp.meta.nvar)
```
Comment (amontoison) [inline src/method.jl:138]:
```suggestion
  gp .= zero(T)
```

**Review by amontoison — CHANGES_REQUESTED** [review]

File: src/method.jl (hunk 1 above)
Comment (tmigot) [reply src/method.jl]: Thanks! I removed it.

**Comment by amontoison** [comment]
Could you test an in-place `solve!` to check that It works fine when you reuse the workspace?
It could be even better if you also add an allocation test.

**Comment by tmigot** [comment]
> Could you test an in-place `solve!` to check that It works fine when you reuse the workspace? It could be even better if you also add an allocation test.

Yes, good idea. I will add this to the todo-list here: https://github.com/JuliaSmoothOptimizers/Percival.jl/issues/101 
This is a preliminary modification.

**Review by amontoison — APPROVED** [review]

File: src/method.jl [hunk 4]
```diff
@@ -70,8 +75,36 @@ Implementation of an augmented Lagrangian method. The following keyword paramete
...
+  return PercivalSolver{V}(x, gx, gL, gp, Jtv)
+end
+
+@doc (@doc PercivalSolver) function percival(
+  ::Val{:equ},
+  nlp::AbstractNLPModel;
+  kwargs...
+)
+  solver = PercivalSolver(nlp)
+  return solve!(Val(:equ), solver, nlp; kwargs...)
```
Comment (dpo) [inline src/method.jl:102]:
```suggestion
  solve!(Val(:equ), solver, nlp; kwargs...)
```

File: src/method.jl [hunk 5]
```diff
@@ -58,6 +58,11 @@ end
 """
     percival(nlp)
 
+---
+
```
Comment (dpo) [inline src/method.jl:62]: remove blank lines and `---`.

File: src/method.jl [hunk 6]
```diff
@@ -172,15 +209,18 @@ function percival(
...
     project_step!(gp, al_nlp.x, -gL, nlp.meta.lvar, nlp.meta.uvar) # Proj(x - gL) - x
     normgp = norm(gp)
 
     iter += 1
     el_time = time() - start_time
     rem_eval = max_eval - neval_obj(nlp)
     solved = normgp ≤ ϵd && normcx ≤ ϵp
-    infeasible = al_nlp.μ > 1e16 && norm(jtprod(nlp, al_nlp.x, al_nlp.cx)) < √ϵp * normcx
+    jtprod!(nlp, al_nlp.x, al_nlp.cx, solver.Jtv)
+    infeasible = al_nlp.μ > 1e16 && norm(solver.Jtv) < √ϵp * normcx
```
Comment (dpo) [inline src/method.jl:223]: 1e16 should be `1/eps(T)` here.

File: src/method.jl [hunk 7]
```diff
@@ -172,15 +209,18 @@ function percival(
...
     tired = iter > max_iter || el_time > max_time || neval_obj(nlp) > max_eval || al_nlp.μ > 1e16
```
Comment (dpo) [inline src/method.jl:224]: replace last condition with `infeasible`?

**Review by dpo — CHANGES_REQUESTED** [review]

File: src/method.jl (hunk 7 above)
Comment (tmigot) [reply src/method.jl]: That's true. We could add ` stalled` boolean for that. What do you think?

Comment (dpo) [reply src/method.jl]: yes, or something more explicit like `penalty too large`.

File: src/method.jl [hunk 8]
```diff
@@ -214,9 +214,9 @@ function solve!(
     rem_eval = max_eval - neval_obj(nlp)
     solved = normgp ≤ ϵd && normcx ≤ ϵp
     jtprod!(nlp, al_nlp.x, al_nlp.cx, solver.Jtv)
-    infeasible = al_nlp.μ > 1 / eps(T) && norm(solver.Jtv) < √ϵp * normcx
-    tired =
-      iter > max_iter || el_time > max_time || neval_obj(nlp) > max_eval || al_nlp.μ > 1 / eps(T)
+    penalty_too_large = al_nlp.μ > 1 / eps(T)
+    infeasible = penalty_too_large && norm(solver.Jtv) < √ϵp * normcx
+    tired = iter > max_iter || el_time > max_time || neval_obj(nlp) > max_eval
```
Comment (dpo) [inline src/method.jl:219]: or penalty too large?

Comment (tmigot) [reply src/method.jl:219]: What do you mean? So, I get rid of `infeasible`?

Comment (dpo) [reply src/method.jl:219]: nvm

**Review by dpo — APPROVED** [review]

**Pull request merged**